src/filesystem/impls/filer/lib/ImageResizer.js

```diff
@@ -0,0 +1,137 @@
+
+    // 1KB
+    var KB = 1024;
+    // Ideal size for images
+    var TARGET_SIZE_KB = 250 * KB;
+    // Resize error tolerance
+    var TOLERANCE_KB = 20 * KB;
+    // Max number of resize passes we allow
+    var MAX_RESIZE_PASSES = 5;
+
+
+    function ImageResizer(path, data) {
+        this.path = decodePath(path);
+        this.type = Content.mimeFromExt(Path.extname(this.path));
+        this.size = data.length;
+        this.blob = new Blob([data], {type: this.type});
+        this.url = URL.createObjectURL(this.blob);
+        this.canvas = document.createElement('canvas');
+    }
+    ImageResizer.prototype.cleanup = function() {
+        URL.revokeObjectURL(this.url);
+    };
+    ImageResizer.prototype.resizedToBuffer = function(callback) {
+        var canvas = this.canvas;
+
+        // If possible, use native methods to read out canvas data
+        if (canvas.toBlob) {
+            canvas.toBlob(function(blob) {
+                var fileReader = new FileReader();
+                fileReader.onload = function() {
+                    var buffer = new Buffer(this.result);
+                    callback(null, buffer);
+                };
+                fileReader.onerror = function() {
+                    callback(new Error("unable to read data"));
+                };
+                fileReader.readAsArrayBuffer(blob);
+            });
+        }
+        // Fallback for browsers without necessary native methods
+        else {
+            var base64Str = canvas.toDataURL(self.type).split(',')[1];
+            var buffer = base64ToBuffer(base64Str);
+            callback(null, buffer);
+        }
+    };
+    ImageResizer.prototype.resize = function(callback) {
+        var self = this;
+        var buffer;
+        var passes = 0;
+        // do a "best guess" initial scale
+        var scale = Math.sqrt(TARGET_SIZE_KB / self.size);
+        var step = scale / 2;
+
+        function finish(err, buffer) {
+            self.cleanup();
+            callback(err, buffer);
+        }
+
+        function resizePass() {
+            console.log("resizePass", passes, "scale", scale);
+
+            if(passes++ > MAX_RESIZE_PASSES) {
+                return finish(null, buffer);
+            }
+
+            self.canvas.width = img.width * scale;
+            self.canvas.height = img.height * scale;
+
+            console.log("width", self.canvas.width, "height", self.canvas.height);
+
+            pica.resize(img, self.canvas)
+            .then(function() {
+                self.resizedToBuffer(function(err, buffer) {
+                    if(err) {
+                        return finish(err);
+                    }
+
+                    // Too big?
+                    if(buffer.length > TARGET_SIZE_KB + TOLERANCE_KB) {
+                        console.log("Resized image too big", buffer.length);
+                        scale = scale - step;
+                        step /= 2;
+                        resizePass();
```

Comment:
It is always best to avoid recursion when possible, which for an algorithm like this is certainly the case. Can you explain why this is a nesting call rather than a `while` loop now?